### PR TITLE
Cleanup init_process_group

### PR DESCRIPTION
**Pull request by kuttas**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19375 Fix eddp_trainer_test regression, rewrite test to be less flaky&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14980467/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19033 Cleanup init_process_group**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14813070/)

torch.distributed.init_process_group() has had many parameters added, but the contract isn't clear. Adding documentation, asserts, and explicit args should make this clearer to callers and more strictly enforced.

Differential Revision: [D14813070](https://our.internmc.facebook.com/intern/diff/D14813070/)

### CODE REVIEW

**Comment by kuttas** [comment]
The one failed test (https://circleci.com/gh/pytorch/pytorch/1230477) looks like it actually succeeded but:

"GitHub rate limited us while adding a commit status of success: You have triggered an abuse detection mechanism. Please wait a few minutes before you try again."

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -347,15 +362,14 @@ def init_process_group(backend,
...
+    assert (store is None) or (init_method is None), \
+        "Cannot specify both init_method and store."
+
     if store is not None:
-        assert world_size > 0, 'world_size needs to be positive'
-        assert rank >= 0, 'rank needs to be non-negative'
-    assert len(kwargs) == 0, \
-        "got unexpected keyword arguments: %s" % ",".join(kwargs.keys())
+        assert world_size > 0, 'world_size must be positive if using store'
+        assert rank >= 0, 'rank must be non-negative if using store'
```
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py]: nit: shall we mention this in the doc?

**Review by mrshenli — APPROVED** [review]
Thanks @kuttas. Yes, the test has passed.

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (pietern) [reply torch/distributed/distributed_c10d.py]: Isn't it safe to think this is implied?

Comment (pietern) [reply torch/distributed/distributed_c10d.py]: Ah, I see these asserts fire if you don't set them at all, and they are set to their default value of `-1`. I think the assert should reflect this and mention that the caller must specify both `world_size` and `rank` keyword arguments to the `init_process_group` call. Then they'll know what to do. In isolation they are not that helpful (if you don't know about those kwargs).

Comment (kuttas) [reply torch/distributed/distributed_c10d.py]: @mrshenli - just want to double check, the comments in this function explain the conditions under which world_size and rank are required (they aren't required if you pass in a URL with them encoded already). If that's good enough I'll resolve this conversation.

Comment (mrshenli) [reply torch/distributed/distributed_c10d.py]: @kuttas yes, that's good enough. :)

**Pull request closed without merging**